docs/source/advanced/contrib_thunder.rst

```diff
@@ -0,0 +1,468 @@
+and so it does not need to be a primitive to enable any functionality.
+Many libraries, including `C++’s standard library <https://en.cppreference.com/w/c/numeric/math/expm1>`_,
+still define an ``expm1`` operation for numerical accuracy, and so ``expm1`` is a
+`primitive <https://github.com/Lightning-AI/lightning-thunder/blob/888b46324462fba70f93d5017bc0d99025f05091/thunder/core/prims.py#L1791>`_ in *Thunder*.
+
+
+-----------------
+The Core Language
+-----------------
+Above the primitives is the ``core`` language, or `clang <https://github.com/Lightning-AI/lightning-thunder/blob/main/thunder/clang/__init__.py>`_.
+Clang operations are mostly written like any other Python operation.
+They ultimately call the primitive operations, although they may call other operations before doing so
+(for example, ``clang.foo`` might call ``clang.bar`` which calls ``prims.bar``).
+
+Core language operations are intended to be common functionality that’s useful when
+defining practitioner-facing language like ``torch`` or ``numpy``.
+Many of these operations are just wrappers around primitive operations.
+For example, the elementwise binary primitives are as simple as possible, so they don’t perform broadcasting or type promotion.
+The core language elementwise binary operations, however, do perform broadcasting and type promotion.
+
+Before adding a core language operation consider if the functionality expressed is universal enough.
+
+As a style note, operations in *Thunder* should defer as much error checking as possible.
+For example, if a primitive’s meta function will perform an error check for ``X``,
+then the core language operation that calls it should generally not also check for ``X``.
+
+
+------------------
+The Torch Language
+------------------
+To translate ``torch`` operations into something that *Thunder* understands we define a
+`torch language <https://github.com/Lightning-AI/lightning-thunder/blob/main/thunder/torch/__init__.py>`_.
+Operations in the ``torch`` language should be written as if a practitioner were writing them,
+and they should reflect the behavior of their corresponding torch operations (small deviations are sometimes OK).
+
+When a program is interpreted, torch operations are remapped into these operations, which ultimately call primitive operations.
+
+
+----------------
+Language Context
+----------------
+In the ``core`` and ``torch`` languages functions are decorated to set a *language context* and – for ``torch`` operations – to describe
+how to map operations like ``torch.foo`` into ``thunder.torch.foo``.
+
+The language context determines what properties and methods tensor objects have.
+For example, when ``a + b`` is written, the language context decides what that addition means.
+Or when ``a.size`` is used, the language context determines what that means (and it’s different in PyTorch and NumPy).
+
+
+---------------------------------------
+Adding operations to the Torch executor
```

Comment:
This header could be given more weights since it marks the "Theory is over, let's look at example" boundary